rust/tvm/src/ir/diagnostics/codespan.rs

```diff
@@ -0,0 +1,216 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+//! A TVM diagnostics renderer which uses the Rust `codespan` library
+//! to produce error messages.
+//!
+//! This is an example of using the exposed API surface of TVM to
+//! customize the compiler behavior.
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+
+use codespan_reporting::diagnostic::{Diagnostic as CDiagnostic, Label, Severity};
+use codespan_reporting::files::SimpleFiles;
+use codespan_reporting::term::termcolor::{ColorChoice, StandardStream};
+use codespan_reporting::term::{self};
+
+use super::*;
+use crate::ir::source_map::*;
+
+/// A representation of a TVM Span as a range of bytes in a file.
+struct ByteRange<FileId> {
+    /// The file in which the range occurs.
+    #[allow(dead_code)]
+    file_id: FileId,
+    /// The range start.
+    start_pos: usize,
+    /// The range end.
+    end_pos: usize,
+}
+
+/// A mapping from Span to ByteRange for a single file.
+enum FileSpanToByteRange {
+    AsciiSource(Vec<usize>),
+    #[allow(dead_code)]
+    Utf8 {
+        /// Map character regions which are larger then 1-byte to length.
+        lengths: HashMap<isize, isize>,
+        /// The source of the program.
+        source: String,
+    },
+}
+
+impl FileSpanToByteRange {
+    /// Construct a span to byte range mapping from the program source.
+    fn new(source: String) -> FileSpanToByteRange {
+        if source.is_ascii() {
+            let line_lengths = source.lines().map(|line| line.len()).collect();
+            FileSpanToByteRange::AsciiSource(line_lengths)
+        } else {
+            panic!()
+        }
+    }
+
+    /// Lookup the corresponding ByteRange for a given Span.
+    fn lookup(&self, span: &Span) -> ByteRange<String> {
+        use FileSpanToByteRange::*;
+
+        let source_name: String = span.source_name.name.as_str().unwrap().into();
+
+        match self {
+            AsciiSource(ref line_lengths) => {
+                let start_pos = (&line_lengths[0..(span.line - 1) as usize])
+                    .into_iter()
+                    .sum::<usize>()
+                    + (span.column) as usize;
+                let end_pos = (&line_lengths[0..(span.end_line - 1) as usize])
+                    .into_iter()
+                    .sum::<usize>()
+                    + (span.end_column) as usize;
+                ByteRange {
+                    file_id: source_name,
+                    start_pos,
+                    end_pos,
+                }
+            }
+            _ => panic!(),
+        }
+    }
+}
+
+/// A mapping for all files in a source map to byte ranges.
+struct SpanToByteRange {
```

Comment:
Is this a place where we could integrating like tracing-style spans? (looking for PR ideas)